modules/transport-netty4/src/test/java/org/elasticsearch/http/netty4/Netty4HttpServerTransportTests.java

```diff
@@ -282,6 +282,52 @@ public void dispatchBadRequest(final RestChannel channel, final ThreadContext th
         assertThat(causeReference.get(), instanceOf(TooLongFrameException.class));
     }
 
+    public void testLargeCompressedResponse() throws InterruptedException {
+        final String responseString = randomAlphaOfLength(4 * 1024 * 1024);
+        final String url = "/thing";
+        final HttpServerTransport.Dispatcher dispatcher = new HttpServerTransport.Dispatcher() {
+
+            @Override
+            public void dispatchRequest(final RestRequest request, final RestChannel channel, final ThreadContext threadContext) {
+                if (url.equals(request.uri())) {
+                    channel.sendResponse(new BytesRestResponse(OK, responseString));
+                } else {
+                    logger.error("--> Unexpected successful uri [{}]", request.uri());
+                    throw new AssertionError();
+                }
+            }
+
+            @Override
+            public void dispatchBadRequest(final RestChannel channel, final ThreadContext threadContext, final Throwable cause) {
+                logger.error(new ParameterizedMessage("--> Unexpected bad request [{}]",
+                    FakeRestRequest.requestToString(channel.request())), cause);
+                throw new AssertionError();
+            }
+
+        };
+
+        try (Netty4HttpServerTransport transport = new Netty4HttpServerTransport(
+            Settings.EMPTY, networkService, bigArrays, threadPool, xContentRegistry(), dispatcher, clusterSettings,
+            new SharedGroupFactory(Settings.EMPTY))) {
+            transport.start();
+            final TransportAddress remoteAddress = randomFrom(transport.boundAddress().boundAddresses());
+
+            try (Netty4HttpClient client = new Netty4HttpClient()) {
+                DefaultFullHttpRequest request = new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, url);
+                request.headers().add(HttpHeaderNames.ACCEPT_ENCODING, randomFrom("deflate", "gzip"));
+                final FullHttpResponse response = client.send(remoteAddress.address(), request);
+                try {
+                    assertThat(response.status(), equalTo(HttpResponseStatus.OK));
```

Comment:
There is no reliable way to determine if the response is chunked or not that does not involve textually decoding HTTP/1.1. That is completely handled by Netty. And through the pipeline Netty changes that header several times. (If it is content-length on the wire, Netty deletes that and changes it to chunked in the object decoder. It changes it back to content-length after the aggregation is complete.) Both transfer-encodings are valid so Netty abstracts it away.

In terms of preventing large allocations - I can do a bunch of casting if you want. I would need to cast the allocator to `NoDirectBuffers` (and make that inner class public) extract the delegate and iff it is pooled, then iterate through the arenas and see if there have been any large allocations. However, this allocator is static and used by both the testing and production code so it will be brittle to a concurrent failure where test code does a huge allocation. And a lot of the time this allocator will be unpooled based on the heap size of tests. But that seems to me to be the only way to verify.